cheat.py

```diff
@@ -604,35 +623,90 @@ def print_planet(self, planet):
+                    next_heal = randint(120,180)
+                    #boss_fails_allowed = 10
+                    next_loop = time()
+                    boss_hp = 10000000
+                    #Loop until break
+                    while(True):
+                        #reset heal on each iteration
+                        heal = 0
+                        #submit every 5 seconds
+                        if((next_loop+5)<time()):
+                            next_loop = time()
+                            #send boss damage
+                            full_response = game.report_boss_damage(heal);
+                            #on E11, restart, bugged for now, fix it later
+                            #if 'headers' in full_response and int(full_response.headers.get('X-eresult', -1)) == 11:
+                            #    game.log("Got invalid state. Restarting")
+                            #    deadline = 0
+                            #    break;
+                            response = full_response['response']
+                            #If there is a battle complete field
+                            if(response.get('game_over') and response.get('game_over')!="" or boss_hp == 0):
+                                game.log("Boss Battle Completed")
+                                break
+                            #If waiting for players, wait a couple seconds, then try again
+                            if(response.get('waiting_for_players') and response.get('waiting_for_players') != ""):
+                                game.log("Waiting for Players")
+                                #sleep to avoid DDOSDB
+                                sleep(2)
+                                continue
+                            #You're in a boss battle at this point, send data
+                            else:
+                                heal = 0
+                                #do healing after between 120 and 180 seconds
+                                if(time_last_heal + next_heal < time()):
+                                    time_last_heal = time()
+                                    next_heal = 120
+                                    heal = 1
+                                if('boss_status' in response):
+                                    if(boss_max_hp == 1):
+                                        boss_max_hp = response['boss_status']['boss_max_hp']
+                                    boss_hp = response['boss_status']['boss_hp']
+                                    #Print out each player.
+                                    for player in response['boss_status']['boss_players']:
+                                        #if(player['accountid'] == game.account_id):
+                                        game.log(str(boss_hp) + " / " + str(boss_max_hp) + " Player: " + re.sub(r'[^\x00-\x7f]',r'', player['name']) + ' Player HP ' + str(player['hp']) + '/' + str(player['max_hp']) + " XP Gained: " + str(player['xp_earned']) )
```

Comment:
1. `log` function supports `printf`-like formatting, see how it's used in a code;
2. Regexp is incorrect, it should be `[^\x20-\x7e]+`;
2.1. Compile regexp during script init and use compiled one.